crypto/http/http_client.c

```diff
@@ -0,0 +1,1200 @@
+            OPENSSL_free(current_url);
+            current_url = redirection_url;
+            goto redirected;
+        }
+        OPENSSL_free(redirection_url);
+    }
+
+ end:
+    OPENSSL_free(current_url);
+    return resp;
+}
+
+/* Get ASN.1-encoded data via HTTP from server at given URL */
+ASN1_VALUE *OSSL_HTTP_get_asn1(const char *url,
+                               const char *proxy, const char *proxy_port,
+                               OSSL_HTTP_bio_cb_t bio_update_fn, void *arg,
+                               const STACK_OF(CONF_VALUE) *headers,
+                               int maxline, unsigned long max_resp_len,
+                               long timeout, const char *expected_content_type,
+                               const ASN1_ITEM *it)
+{
+    BIO *mem;
+    ASN1_VALUE *resp = NULL;
+
+    if (url == NULL || it == NULL) {
+        HTTPerr(0, ERR_R_PASSED_NULL_PARAMETER);
+        return NULL;
+    }
+    if ((mem = OSSL_HTTP_get(url, proxy, proxy_port, bio_update_fn, arg,
+                             headers, maxline, max_resp_len, timeout,
+                             expected_content_type, 1 /* expect_asn1 */))
+         != NULL)
+        resp = BIO_mem_d2i(mem, it);
+    BIO_free(mem);
+    return resp;
+}
+
+/* Post ASN.1-encoded request via HTTP to server return ASN.1 response */
+ASN1_VALUE *OSSL_HTTP_post_asn1(const char *server, const char *port,
+                                const char *path, int use_ssl,
+                                const char *proxy, const char *proxy_port,
+                                OSSL_HTTP_bio_cb_t bio_update_fn, void *arg,
+                                const STACK_OF(CONF_VALUE) *headers,
+                                const char *content_type,
+                                ASN1_VALUE *req, const ASN1_ITEM *req_it,
+                                int maxline, unsigned long max_resp_len,
+                                long timeout, const char *expected_ct,
+                                const ASN1_ITEM *rsp_it)
+{
+    BIO *req_mem;
+    BIO *res_mem;
+    ASN1_VALUE *resp = NULL;
+
+    if (req == NULL) {
+        HTTPerr(0, ERR_R_PASSED_NULL_PARAMETER);
+        return NULL;
+    }
+    /* remaining parameters are checked indirectly */
+
+    req_mem = HTTP_ASN1_item2BIO(req_it, req);
+    res_mem = OSSL_HTTP_transfer(server, port, path, use_ssl, proxy, proxy_port,
+                                 bio_update_fn, arg, headers, content_type,
+                                 req_mem /* may be NULL */, maxline,
+                                 max_resp_len, timeout,
+                                 expected_ct, 1 /* expect_asn1 */, NULL);
+    BIO_free(req_mem);
+    if (res_mem != NULL)
+        resp = BIO_mem_d2i(res_mem, rsp_it);
+    BIO_free(res_mem);
+    return resp;
+}
+
+/* BASE64 encoder used for encoding basic proxy authentication credentials */
+static char *base64encode(const void *buf, size_t len)
+{
+    int i;
```

Comment:
No, because `EVP_EncodeBlock()` has type `int` and there is a comparison `i < 0`.
An actual improvement in this function is to change the type casts in `ossl_assert(i <= (int)outl)`, which I've done now.